### PR TITLE
Feature/digest client

### LINKED ISSUE #3588: Digest Auth client middleware
Http4s provide Digest auth middleware to implement server, but luck of client middleware
This can be really not so trivial task for somebody who will just start working with http4s. 
I have this already being implemented, need to revisit and contribute this to become a part of http4s project

### PR DESCRIPTION
**Pull request by FrancescoSerra**
Closes #3588

Follows up on PR #3619

I'm not overly fond of this implementation, so I'd like to get feedback - I kept the changes to the minimum

### CODE REVIEW

File: client/shared/src/main/scala/org/http4s/client/digest/NonceCounter.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+object NonceCounter {
+  val MaxNonce: Long = java.lang.Long.parseLong("FFFFFFFF", 16)
+
+  def make[F[_]: Async](
+      ttl: FiniteDuration = 1.minute,
+      maxNonce: Long = MaxNonce,
+  ): F[NonceCounter[F]] =
+    Random
+      .javaSecuritySecureRandom[F]
```
Comment (FrancescoSerra) [inline client/shared/src/main/scala/org/http4s/client/digest/NonceCounter.scala:45]: According to the relative [RFC](https://datatracker.ietf.org/doc/html/rfc7616), there's no strict need for a cryptographically secure cnonce - however I think this won't hurt.

Comment (armanbilge) [reply client/shared/src/main/scala/org/http4s/client/digest/NonceCounter.scala:45]: Is this the relevant section?
https://datatracker.ietf.org/doc/html/rfc7616#section-5.12

> The security of this protocol is critically dependent on the
   randomness of the randomly chosen parameters, such as client and
   server nonces.  These should be generated by a strong random or
   properly seeded pseudorandom source (see [[RFC4086](https://datatracker.ietf.org/doc/html/rfc4086)]).

Comment (FrancescoSerra) [reply client/shared/src/main/scala/org/http4s/client/digest/NonceCounter.scala:45]: yes, I think the `properly seeded pseudorandom` seems covered by the standard `Random` typeclass.

Comment (FrancescoSerra) [reply client/shared/src/main/scala/org/http4s/client/digest/NonceCounter.scala:45]: Same! I'm only interpreting the RFC text, but I'm no expert and I might be blatantly wrong 😊

**Pull request closed without merging**